lib/Dialect/SV/Transforms/HWMemSimImpl.cpp

```diff
@@ -190,6 +253,73 @@ void HWMemSimImplPass::runOnOperation() {
     if (genOp.descriptor() == "FIRRTL_Memory") {
       auto mem = analyzeMemOp(oldModule);
 
+      // If this memory has defined write-after-write behavior, we need to
+      // determine which sets of write ports are driven by the same clock.
+      if (mem.writeUnderWrite == 1) {
+
+        auto symbol = SymbolRefAttr::get(oldModule.getNameAttr());
+        auto instances = symbolDigger.get(symbol);
+        bool differentClocks = false;
+        auto buildWriteSets = [&](Operation *op) {
+          llvm::DenseMap<Value, unsigned> clockToLeader;
+          llvm::EquivalenceClasses<unsigned> writeSets;
+          for (size_t i = mem.numReadPorts * 3 + 0, e = op->getNumOperands();
+               i < e; i += 5) {
+            Value clock = op->getOperand(i);
+            auto result = clockToLeader.insert({clock, i});
+            // Clock not known
+            if (result.second)
+              writeSets.insert(i);
+            // Clock already known
+            else
+              writeSets.unionSets(result.first->second, i);
+          }
+          return writeSets;
+        };
+
+        // Loop over each instance of the memory.  Save the write port mapping
```

Comment:
I would embed the clock information in the memory schema and do the analysis when mem is lowered to a generator node.  The Lowering should be able to go just from the schema with no analysis, thus there is a 1:1 relation from generator node to memory impl.

Reply:
That makes sense. This would also would, I think, work around the problem of memories getting incorrectly deduplicated because the analysis would happen during schema generation.

Any idea on how to encode the disjoint set clock information into an attribute? An arrayattr of clock ids?

Reply:
Updated to do it this way in the rebase. Much cleaner. Great suggestion.